staging/src/k8s.io/apiserver/pkg/server/openapi/openapi.go

```diff
@@ -64,14 +94,37 @@ func RegisterOpenAPIService(servePath string, webServices []*restful.WebService,
 		return err
 	}
 
-	mux.UnlistedHandleFunc(servePath, func(w http.ResponseWriter, r *http.Request) {
-		resp := restful.NewResponse(w)
-		if r.URL.Path != servePath {
-			resp.WriteErrorString(http.StatusNotFound, "Path not found!")
-		}
-		// TODO: we can cache json string and return it here.
-		resp.WriteAsJson(o.swagger)
-	})
+	mime.AddExtensionType(".json", MIME_JSON)
+	mime.AddExtensionType(".pb", MIME_PB)
+	mime.AddExtensionType(".gz", MIME_PB_GZ)
+
+	type fileInfo struct {
+		ext  string
+		data []byte
+	}
+
+	files := []fileInfo{
+		{".json", o.swaggerBytes},
+		{".pb", o.swaggerPb},
+		{".pb.gz", o.swaggerPbGz},
+	}
+
+	for _, file := range files {
+		path := servePathBase + file.ext
```

Comment:
I guess I would have expected a single endpoint that used the Accept: header to determine the content. I suppose this works, too. I am very uncomfortable with the fact that clients aren't sending anything indicating the version of the proto schema that they want back, since proto is not self describing. A schema version in the filename, or via e.g. the Accept: header, would make me feel a lot better.